File: Question/30_Substring_with_Concatenation_of_All_Words.cpp
//
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        if(s.empty() || words.empty()) return {};
        unordered_map<string, int> word_map;
        for(auto& word : words) ++word_map[word];

        int len = words[0].size();
        int n = words.size();
        vector<int> ans;
        for(int i = 0; i <= (int)s.size() - n * len; ++i) {
            auto count_map = word_map;
            int j = i;
            while(j <= i + n*len) {
                string sub_str = s.substr(j, len);
                if(!count_map.count(sub_str) || count_map[sub_str] == 0)
                    break;
                --count_map[sub_str];
                j += len;
            }
            if(j == (i + n*len))
                ans.push_back(i);
        }
        return ans;
    }
};

//sliding window
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        if(s.empty() || words.empty()) return {};
        unordered_map<string, int> word_map;
        for(auto& word : words) ++word_map[word];

        int len = words[0].size();
        int n = words.size();
        vector<int> ans;
        for(int i = 0; i < len; ++i) {
            auto counts = word_map;
            int num = n;
            for(int j = i; j + len <= s.size(); j += len) {
                string cur = s.substr(j, len);
                if(counts[cur]-- > 0) --num;
                if(j - n * len >= 0) { //out of current window
                    string out = s.substr(j - n * len, len);
                    if(++counts[out] > 0) ++num; //回收左窗口
                }
                if(num == 0) ans.push_back(j - (n - 1) * len);
            }
        }
        return ans;
    }
};

File: Question_Java/30_Substring_with_Concatenation_of_All_Words.java
// brute force
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        HashMap<String, Integer> mem = new HashMap<>();
        for (String word : words) {
            Integer value = mem.getOrDefault(word, 0);
            mem.put(word, ++value);
        }

        int len = words[0].length();
        int n = words.length;
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i <= s.length() - n * len; ++i) {
            HashMap<String, Integer> tmem = (HashMap)mem.clone();
            int j = i;
            for ( ; j < i + n * len; j += len) {
                String sub_str = s.substring(j, j + len);
                if (!tmem.containsKey(sub_str) || tmem.get(sub_str) == 0) {
                    break;
                }
                int value = tmem.get(sub_str);
                tmem.put(sub_str, --value);
            }
            if (j == i + n * len) ans.add(i);
        }
        return ans;
    }
}

// sliding window
